.github/scripts/check_imports/check_imports.py

```diff
@@ -0,0 +1,222 @@
+        """Instantiate configuration from a JSON file if it exists."""
+        if not path.exists():
+            return cls()
+        with path.open("r", encoding="utf-8") as handle:
+            data = json.load(handle)
+        modules = data.get("modules", [])
+        path_scoped = data.get("files", {})
+        return cls(modules, path_scoped)
+
+    def is_allowed(self, module: str, file_path: Path) -> bool:
+        """Return True when a module is allow-listed for the given file path."""
+        canonical_module = canonicalize_module_name(module)
+        if canonical_module in self.module_allowlist:
+            return True
+        resolved = file_path
+        modules = self.path_scoped_allowlist.get(resolved)
+        if modules:
+            return canonical_module in modules
+        for parent in resolved.parents:
+            modules = self.path_scoped_allowlist.get(parent)
+            if modules and canonical_module in modules:
+                return True
+        return False
+
+
+def canonicalize_module_name(name: str) -> str:
+    """Return the top-level portion of a dotted module path."""
+    return name.split(".")[0]
+
+
+def discover_python_files(paths: Sequence[str], restrict_to: Optional[Sequence[str]] = None) -> list[Path]:
+    """Collect Python files from individual files or by walking directories."""
+    python_files: list[Path] = []
+    restricted_paths: Optional[list[Path]] = None
+    if restrict_to:
+        restricted_paths = [Path(p).resolve() for p in restrict_to]
+
+    for raw_path in paths:
+        path = Path(raw_path)
+        # Skip hidden files/directories in any part of the path
+        if any(part.startswith(".") for part in path.parts):
+            continue
+        if path.is_file() and path.suffix == ".py":
+            python_files.append(path)
+        elif path.is_dir():
+            for candidate in path.rglob("*.py"):
+                if any(part.startswith(".") for part in candidate.parts):
+                    continue
+                python_files.append(candidate)
+
+    # Filter files to only those within restricted directories
+    if restricted_paths:
+        filtered_files: list[Path] = []
+        for file_path in python_files:
+            resolved = file_path.resolve()
+            for restricted in restricted_paths:
+                try:
+                    resolved.relative_to(restricted)
+                    filtered_files.append(file_path)
+                    break
+                except ValueError:
+                    # File is not within this restricted path
+                    continue
+        return filtered_files
+
+    return python_files
+
+
+@lru_cache(maxsize=1)
+def build_stdlib_index() -> frozenset[str]:
+    """Return a set containing names of standard-library modules."""
+    candidates: set[str] = set(sys.builtin_module_names)
+    candidates.update(canonicalize_module_name(name) for name in sys.stdlib_module_names)
```

Comment:
Is this needed or can we just directly leverage `sys.stdlib_module_names`?

Reply:
Yes..sys.stdlib_module_names doesn’t cover every standard-library module, so relying on it alone would produce false positives. We need to include sys.builtin_module_names as well, and canonicalize the names, to ensure the full stdlib set is captured.